text/0497-glue-l2-construct.md

```diff
@@ -0,0 +1,434 @@
+to schedule and run tasks that don't require an Apache Spark environment.
+
+We’ll default to `PythonVersion.3_9`. Python shell jobs have a MaxCapacity feature.
+Developers can choose MaxCapacity = `0.0625` or MaxCapacity = `1`. By default,
+MaxCapacity will be set `0.0625`. Python 3.9 supports preloaded analytics
+libraries using the `library-set=analytics` flag, and this feature will
+be enabled by default.
+
+```
+new glue.PythonShellJob(this, 'PythonShellJob', {
+   script: glue.Code.fromBucket('bucket-name', 's3prefix/path-to-python-script'),
+});
+```
+
+Optional overrides:
+
+```
+new glue.PythonShellJob(this, 'PythonShellJob', {
+    glueVersion: glue.GlueVersion.V1_0,
+    pythonVersion: glue.PythonVersion.3_9,
+    script: glue.Code.fromBucket('bucket-name', 's3prefix/path-to-python-script'),
+    description: 'an example Python Shell job',
+    numberOfWorkers: 20,
+});
+```
+
+### Ray Jobs
+
+Glue ray only supports worker type Z.2X and Glue version 4.0. Runtime
+will default to `Ray2.3` and min workers will default to 3.
+
+```
+new glue.GlueRayJob(this, 'GlueRayJob', {
+    script: glue.Code.fromBucket('bucket-name', 's3prefix/path-to-python-script'),
+});
+```
+
+Developers can override min workers and other Glue job fields
+
+```
+new glue.GlueRayJob(this, 'GlueRayJob', {
+  runtime: glue.Runtime.RAY_2_2,
+  script: glue.Code.fromBucket('bucket-name', 's3prefix/path-to-python-script'),
+  numberOfWorkers: 50
+});
+```
+
+### Required, Optional, and Overridable Parameters
+
+Each of these parameters are documented in [Glue's public documentation](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api.html);
+this table lists the parameters that will or will not be required, optional,
+and/or overridable for the Glue L2 construct.
+
+|Parameter|Required|Default Value|Overridable|
+|:---|:---|:---|:---|
+|ScriptLocation|Yes: S3 location|CommandName|yes S3 location|
+|NumberOfWorkers|No|10|Yes for ETL, STREAMING, RAY jobs|
+|Name|Yes|Default: Auto Generated name|Yes|
+|Role|Yes|None|Yes|
+|Description|No|None|Yes|
+|MaxConcurrentRuns|No|1|Yes|
+|DefaultArguments|No|None|Yes|
+|Connections|No|None|Yes|
+|MaxRetries|No|0|Yes|
+|Timeout|No|2 Days for non-streaming jobs|Yes|
+|SecurityConfiguration|No|None|Yes| |Tags|No|None|Yes|
+|GlueVersion|No|Default: 3.0 for ETL, RAY: 4.0|Yes|
+|Command name|No|ETL, PythonShell, Streaming and GlueRay|No|
+|Command Runtime|No|GlueRay: Ray2.4|No|
+|NonOverridableArguments|No|None|No|
+|ExecutionClass|No|STANDARD / FLEX|No|
+|Command PythonVersion|No|3 for ETL/Streaming, 3.9 for PythonShell|
+|LogUri|No|None|This option is not widely used|
+|MaxCapacity (Python Shell Jobs only)|No|0.0625|Yes|
+
+### Uploading scripts from the same repo to S3
+
+Similar to other L2 constructs, the Glue L2 will automate uploading / updating
+scripts to S3 via an optional fromAsset parameter pointing to a script
+in the local file structure. Developers will provide an existing S3 bucket and
```

Comment:
> Developers will provide an existing S3 bucket and the path to which they'd like the script to be uploaded.

Why do we need this? The whole point of using `fromAsset()` is to hide this implementation detail from the user.

Reply:
Originally there was a question of whether we would support uploading scripts from a local file to S3 or only support existing files in S3. We decided to support both, and all of the other examples show the latter. This section specifically calls out that we'll support the former as well. However, it's not something we can hide from the developer because they have to specify where the file is coming from using fromAsset (local) or fromBucket (already in S3).

Reply:
Sure, but these methods should have different signatures:

```ts
fromAsset('/path/to/file') // <-- user specifies the local file, but the bucket is decided by the CDK
fromBucket(bucketName, objectKey)
```

Reply:
Having two separate methods won't hide the implementation detail from the user, they still have to specify the path or bucket attributes, and they'll have to remember to use a different method to create the job as well. We're following the pattern of Lambda.